### PR TITLE
[RaySGD] Convert the head worker to a local model

### PR DESCRIPTION
**Pull request by maximsmol**
## Why are these changes needed?

Running a worker on head (locally, not as a Ray actor) allows for easier handling of stateful stuff like logging and for easier debugging.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

**Comment by maximsmol** [comment]
CC: @richardliaw

File: python/ray/util/sgd/torch/distributed_torch_runner.py [hunk 1]
```diff
@@ -140,12 +149,37 @@ def _set_model_state_dicts(self, model_state_dicts):
...
+        dist.destroy_process_group()
+        super(DistributedTorchRunner, self).shutdown()
+
+
+class _DummyActor:
+    def cuda_devices(self):
+        return os.environ["CUDA_VISIBLE_DEVICES"]
+
+
+class LocalDistributedRunner(DistributedTorchRunner):
```
Comment (richardliaw) [inline python/ray/util/sgd/torch/distributed_torch_runner.py]:
```suggestion
class LocalDistributedRunner(DistributedTorchRunner):
    """A wrapper for running a distributed Runner on the driver.
    
    A dummy actor is used to reserve resources on the driver node.
    """
```

File: python/ray/util/sgd/torch/examples/sgd-development.yaml [hunk 2]
```diff
@@ -65,15 +69,15 @@ setup_commands:
...
+    - pip install -U ipdb torch torchvision tqdm
     # Install Apex
     - rm -rf apex || true
     - git clone https://github.com/NVIDIA/apex && cd apex && pip install -v --no-cache-dir  ./ || true
 
 
 file_mounts: {
     # This should point to ray/python/ray/util/sgd.
-    # ~/sgd: ../../../sgd,
+    ~/anaconda3/lib/python3.6/site-packages/ray/util/sgd: ../../../sgd,
```
Comment (richardliaw) [inline python/ray/util/sgd/torch/examples/sgd-development.yaml]: let's revert this file?

File: python/ray/util/sgd/torch/torch_trainer.py [hunk 3]
```diff
@@ -256,68 +253,80 @@ def _start_workers(self, num_workers):
...
+            ray.get(remote_setups)
+
+            remote_reporter_sets = [
                 w.set_reporters.remote(
                     [h.create_reporter() for h in self.handlers])
-                for w in self.workers
-            ])
+                for w in self.remote_workers
+            ]
+            self.local_worker.set_reporters(
```
Comment (richardliaw) [inline python/ray/util/sgd/torch/torch_trainer.py]: are we reverting the reporter/handler stuff?

File: python/ray/util/sgd/torch/torch_trainer.py [hunk 4]
```diff
@@ -385,15 +394,15 @@ def train(self,
                 break
             else:
                 self._num_failures += 1
+            h.prepare_for_retry()
```
Comment (richardliaw) [inline python/ray/util/sgd/torch/torch_trainer.py]: if we can get rid of handlers this won't be needed

File: python/ray/util/sgd/torch/torch_trainer.py [hunk 5]
```diff
@@ -413,42 +422,24 @@ def _process_stats(self, worker_stats):
...
-                loop.run_until_complete(asyncio.wait(futures))
-                loop.close()
-
-            return True, worker_trains
-        except RayActorError as exc:
-            logger.exception(str(exc))
-        return False, worker_trains
+            local_worker_stats = self.local_worker.train_epoch(**params)
+        except RuntimeError as err:
+            print(err)
```
Comment (richardliaw) [inline python/ray/util/sgd/torch/torch_trainer.py]: use logger instead?

File: python/ray/util/sgd/torch/examples/sgd-development.yaml (hunk 2 above)
Comment (maximsmol) [reply python/ray/util/sgd/torch/examples/sgd-development.yaml]: yup, that change just slipped through

**Review by richardliaw — APPROVED** [review]
lgtm, can merge when tests pass

**Pull request merged**